enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+2. Do not send any service traffic out of eth0, instead all service traffic goes to eth1. ovn_cluster_router_&lt;unique network> is configured to send service CIDR traffic to ovn-k8s-mp0-&lt;unique network>. ipTables rules in the host only permit access to KAPI/DNS and drop all other service traffic coming from ovn-k8s-mp0-&lt;unique network>. The traffic then gets routed to br-ex and default GR where it hits the service.
+
+While the second option is more complex, it allows for not configuring routes to service addresses in the pod that could hypothetically change.
+
+
+#### Egress IP
+
+This feature works today by labeling and choosing a node+network to be egress, and then OVN logical routes and logical route policies are created which steer traffic from a pod towards a specific gateway router (for primary network egress). From there the packets are SNAT’ed by the OVN GR to the egress IP, and sent to br-ex. Egress IP is cluster scoped, but applies to selected namespaces, which will allow us to only apply the SNAT and routes to the GR and OVN topology elements of that network. In the layer 3 case, the current design used today for the default primary network will need some changes. Since Egress IP may be served on multiple namespaces and thus networks, it is possible that there could be a collision as previously mentioned in the Pod Egress section. Therefore, the same solution provided in that section where the GR SNATs to the masquerade subnet must be utilized. However, once the packet arrives in br-ex we will need a way to tell if it was sent from a pod affected by a specific egress IP. To address this, pkt_mark will be used to mark egress IP packets and signify to br-ex which egress IP to SNAT to. An example where the egress IP is 1.1.1.1 that maps to pkt_mark 10 would look something like this:
+
+
+![Egress IP VRF SGW](images/egress-ip-vrf-sgw.svg)
+
+
+For layer 2, egress IP has never been supported before. With the IC design, there is no need to have an ovn_cluster_router and join switch separating the layer 2 switch network (transit switch) from the GR. For non-IC cases this might be necessary, but for OpenShift purposes will only describe the behavior of IC in this proposal. In the layer 2 IC model, GRs per node on a network will all be connected to the layer 2 transit switch:
+
+
+![Egress IP Layer 2](images/egress-ip-l2-primary.svg)
+
+
+In the above diagram, Node 2 is chosen to be the egress IP node for any pods in namespace A. Pod 1 and Pod 2 have default gateway routes to their respective GR on their node. When egress traffic leaves Pod 2, it is sent towards its GR-A on node 2, where it is SNAT’ed to the egress IP and the traffic sent to br-ex. For Pod 1, its traffic is sent to its GR-A on Node 1, where it is then rerouted towards GR-A on Node 2 for egress.
```

Comment:
the GARP for egress ip is an external facing thing isn't it? Like if the EIP moves from node A to node B, now node B GR sends GARP out towards the external network to update external devices that the EIP is now reachable via node B's MAC. We don't use GARP internally for Egress IP AFAIK. The pod always sends its traffic to the default gw, then we use routing (static routes and logical route policy) to get the packet to the right ovn GR, where it will then do the SNAT to EIP.

Reply:
Oh sorry, yes I meant external facing GARPs but I guess this comment pertains to the local gateway EIP support, not the shared gateway one.